minifi/minifi-c2/minifi-c2-service/src/main/java/org/apache/nifi/minifi/c2/service/ConfigService.java

```diff
@@ -147,6 +175,123 @@ protected ConfigurationProviderInfo initContentTypeInfo(List<ConfigurationProvid
         return new ConfigurationProviderInfo(mediaTypeList, contentTypes, null);
     }
 
+    @POST
+    @Path("/heartbeat")
+    @Consumes(MediaType.APPLICATION_JSON)
+    @Produces(MediaType.APPLICATION_JSON)
+    @ApiOperation(
+            value = "An endpoint for a MiNiFi Agent to send a heartbeat to the C2 server",
+            response = C2HeartbeatResponse.class
+    )
+    @ApiResponses({
+            @ApiResponse(code = 400, message = MESSAGE_400)})
+    public Response heartbeat(
+            @Context HttpServletRequest request, @Context HttpHeaders httpHeaders, @Context UriInfo uriInfo,
+            @ApiParam(required = true) final C2Heartbeat heartbeat) {
+
+        try {
+            authorizer.authorize(SecurityContextHolder.getContext().getAuthentication(), uriInfo);
+        } catch (AuthorizationException e) {
+            logger.warn(HttpRequestUtil.getClientString(request) + " not authorized to access " + uriInfo, e);
+            return Response.status(403).build();
+        }
+
+        List<MediaType> acceptValues = httpHeaders.getAcceptableMediaTypes();
+        boolean defaultAccept = false;
+        if (acceptValues.size() == 0) {
+            acceptValues = Collections.singletonList(MediaType.WILDCARD_TYPE);
+            defaultAccept = true;
+        }
+        if (logger.isDebugEnabled()) {
+            StringBuilder builder = new StringBuilder("Handling request from ")
+                    .append(HttpRequestUtil.getClientString(request))
+                    .append(" with Accept");
+            if (defaultAccept) {
+                builder.append(" default value");
+            }
+            builder.append(": ")
+                    .append(acceptValues.stream().map(Object::toString).collect(Collectors.joining(", ")));
+            logger.debug(builder.toString());
+        }
+
+        try {
+            final String flowId;
+            Map<String, List<String>> parameters = Collections.singletonMap("class", Collections.singletonList(heartbeat.getAgentClass()));
+            ConfigurationProviderValue configurationProviderValue = configurationCache.get(new ConfigurationProviderKey(acceptValues, parameters));
+            org.apache.nifi.minifi.c2.api.Configuration configuration = configurationProviderValue.getConfiguration();
+            try (InputStream inputStream = configuration.getInputStream();
```

Comment:
I'm not sure whether I'm reading this right or not but if there is a heartbeat with an unrecognised agent class (there  is no flow definition for it on the server side) shouldn't we handle that more gracefully? I mean with something like 4XX response or at least without exception? Maybe @kevdoran has better insight here.

Reply:
How about a 200 if no agent class is provided and No Content (204) if no flow is available for that agent class?

Reply:
Sounds good to me.